src/mlpack/methods/xgboost/xgb_exact_numeric_split_impl.hpp

```diff
@@ -0,0 +1,94 @@
+/**
+ * @file methods/xgboost/xgb_exact_numeric_split_impl.hpp
+ * @author Rishabh Garg
+ *
+ * Implementation of exact numeric splitter for XGBoost.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_XGBOOST_XGB_EXACT_NUMERIC_SPLIT_IMPL_HPP
+#define MLPACK_METHODS_XGBOOST_XGB_EXACT_NUMERIC_SPLIT_IMPL_HPP
+
+namespace mlpack {
+namespace ensemble {
+
+template<typename LossFunction>
+template<bool UseWeights, typename VecType, typename MatType,
+          typename WeightVecType>
+double XGBExactNumericSplit<LossFunction>::SplitIfBetter(
+    const double bestGain,
+    const VecType& data,
+    const MatType& input,
+    const WeightVecType& weights,
+    const size_t minimumLeafSize,
+    const double minimumGainSplit,
+    double& splitInfo,
+    AuxiliarySplitInfo& aux,
+    LossFunction& lossFunction)
+{
+  // TODO: Filter out and separate the missing data before sorting to decrease
+  // sensitivity towards missing data.
+
+  // Next, sort the data.
+  arma::uvec sortedIndices = arma::sort_index(data);
+  arma::vec sortedGradients(sortedIndices.n_elem);
+  arma::vec sortedHessians(sortedIndices.n_elem);
+  lossFunction.sortGradAndHess(sortedIndices, sortedGradients,
+      sortedHessians);
+
+  // Sanity check: if the first element is the same as the last, we can't split
+  // in this dimension.
+  if (data[sortedIndices[0]] == data[sortedIndices[sortedIndices.n_elem - 1]])
+    return DBL_MAX;
+
+  // Initialize the binary scan. It initialises the index such that the weight
+  // of the left child is higher than the minChildWeight.
+  size_t index = lossFunction.BinaryScanInitialize(sortedGradients,
+      sortedHessians);
+
+  double bestFoundGain = bestGain + minimumGainSplit;
+  bool improved = false;
+  bool endLoop = false;
+  // Forcing the loop to run for one less than the number of data points to
+  // ensure that we have atleast one element in the right child even when
+  // minChildWeight is lesser than or equal to zero.
+  for (; index < data.n_elem - 1; ++index)
+  {
+    lossFunction.BinaryStep(sortedGradients, sortedHessians, index, endLoop);
+
+    // We have to ensure that the minChildWeight condition is held in the right
+    // child too. So, if at any index, that condition doesn't hold true, then
+    // terminate the loop.
+    if (endLoop) break;
```

Comment:
```suggestion
    if (endLoop)
      break;
```
(Tiny style suggestion.)